### PR TITLE
Prevent more unnecessary float boxing, especially in `if` and `match`

### PR DESCRIPTION
**Pull request by vbrankov**
This patch prevents unnecessary boxing in more language constructs.  Especially important are `if` and `match` blocks, for example:

``` ocaml
let reduce dir price ~by =
  match dir with
  | Buy -> price -. by
  | Sell -> price +. by
```

All OCaml tests passed.  The test code that shows that the extra allocations are gone is here:

> https://github.com/vbrankov/ocaml-demos/blob/master/unnecessary-boxing/test.ml

### CODE REVIEW

File: asmcomp/cmmgen.ml [hunk 1]
```diff
@@ -1238,7 +1252,21 @@ let rec is_unboxed_number = function
         | Pbbswap bi -> Boxed_integer bi
         | _ -> No_unboxing
```
Comment (chambart) [inline asmcomp/cmmgen.ml:1255]: Any non returning primitive could return Undetermined. Praise for instance. (I am not certain there are no other one)

**Comment by vbrankov** [comment]
I've added handling of `int32`, `int64` and `nativeint` constants.

**Comment by alainfrisch** [comment]
I'm currently integrating the patch and will merge it shortly.

I've renamed Undetermined to No_result, to make it clear that it really means that the sub-expression can never return a value (and so it is fine to apply any kind of unboing to it).  And I've also changed the code in the Ulet case to avoid emitting code for the let-body when one knows that the let-bound expression cannot return.

**Comment by alainfrisch** [comment]
Applied to trunk, commit 16271.

**Comment by alainfrisch** [comment]
This ticket can be closed.

**Pull request closed without merging**